aggregator/election_mgr.go

```diff
@@ -0,0 +1,544 @@
+		changeRetrier:       changeRetrier,
+		changeVerifyTimeout: opts.ChangeVerifyTimeout(),
+		electionKeyFmt:      opts.ElectionKeyFmt(),
+		leaderService:       opts.LeaderService(),
+		leaderValue:         campaignOpts.LeaderValue(),
+		state:               electionManagerNotOpen,
+		doneCh:              make(chan struct{}),
+		electionState:       FollowerState,
+		goalStateWatchable:  xwatch.NewWatchable(),
+		resignCh:            make(chan resignAction, 1),
+		sleepFn:             time.Sleep,
+		metrics:             newElectionManagerMetrics(instrumentOpts.MetricsScope()),
+	}
+}
+
+func (mgr *electionManager) Open(shardSetID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerNotOpen {
+		return errElectionManagerExpectNotOpen
+	}
+	mgr.state = electionManagerOpen
+	mgr.electionKey = fmt.Sprintf(mgr.electionKeyFmt, shardSetID)
+
+	_, stateChangeWatch, err := mgr.goalStateWatchable.Watch()
+	if err != nil {
+		return err
+	}
+	go mgr.watchStateChanges(stateChangeWatch)
+
+	_, verifyWatch, err := mgr.goalStateWatchable.Watch()
+	if err != nil {
+		return err
+	}
+	go mgr.verifyPendingFollower(verifyWatch)
+
+	go mgr.startElectionLoop()
+	go mgr.reportMetrics()
+	return nil
+}
+
+func (mgr *electionManager) ElectionState() ElectionState {
+	mgr.RLock()
+	electionState := mgr.electionState
+	mgr.RUnlock()
+	return electionState
+}
+
+func (mgr *electionManager) Resign(ctx context.Context) error {
+	_, watch, err := mgr.goalStateWatchable.Watch()
+	if err != nil {
+		return fmt.Errorf("error creating watch when resigning: %v", err)
+	}
+	defer watch.Close()
+
+	resignAction := resignAction{errCh: make(chan error, 1)}
+	select {
+	case mgr.resignCh <- resignAction:
+	default:
+		return errResignActionAlreadyEnqueued
+	}
+
+	for {
+		select {
+		case <-watch.C():
+			if goalState := watch.Get().(goalState); goalState.state == FollowerState {
+				return nil
+			}
+		case <-ctx.Done():
+			mgr.metrics.resignTimeout.Inc(1)
+			mgr.logError("resign error", ctx.Err())
+			return ctx.Err()
+		case err := <-resignAction.errCh:
+			mgr.metrics.resignErrors.Inc(1)
+			mgr.logError("resign error", err)
+			return err
+		}
+	}
+}
+
+func (mgr *electionManager) Close() error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerOpen {
+		return errElectionManagerExpectOpen
+	}
+	close(mgr.doneCh)
+	mgr.state = electionManagerClosed
+	return nil
+}
+
+func (mgr *electionManager) watchStateChanges(watch xwatch.Watch) {
+	defer watch.Close()
+
+	for {
+		select {
+		case <-mgr.doneCh:
+			return
+		case <-watch.C():
+			mgr.processGoalState(watch.Get().(goalState))
+		case action := <-mgr.resignCh:
+			mgr.resign(action)
+		}
+	}
+}
+
+func (mgr *electionManager) processGoalState(goalState goalState) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	currState := mgr.electionState
+	newState := goalState.state
+	if currState == newState {
+		return
+	}
+	if currState == FollowerState && newState == PendingFollowerState {
```

Comment:
Conceptually it isn't, but because the leader election API currently does not verify states, if the client receives a follower state, it can only treat it as a "pendingFollower" state. Once the leader election logic is adapted to only notify clients of states that have been verified (@mschalle and I just talked offline and he's working on it), this is no longer needed. I'll put a comment here making a note of it.